src/sql-parser/src/parser.rs

```diff
@@ -4207,8 +4219,61 @@ impl<'a> Parser<'a> {
         self.expect_keyword(TABLE)?;
         let if_not_exists = self.parse_if_not_exists()?;
         let table_name = self.parse_item_name()?;
+
         // parse optional column list (schema)
-        let (columns, constraints) = self.parse_columns(Mandatory)?;
+        let (columns, constraints) = match self.parse_columns(Optional)? {
+            Some((c, c1)) => (Some(c), Some(c1)),
+            None => (None, None),
+        };
+        let column_pos = self.peek_prev_pos();
+        let column_next_token = self.peek_token();
+
+        // If this is a read-only table from a source, parse the source and external reference
+        // and ensure that any columns specified do not include full definitions
+        let from_source = if self.parse_keywords(&[FROM, SOURCE]) {
+            let source = self.parse_raw_name()?;
+            self.expect_token(&Token::LParen)?;
+            self.expect_keyword(REFERENCE)?;
+            let _ = self.consume_token(&Token::Eq);
+            let external_reference = self.parse_item_name()?;
+            self.expect_token(&Token::RParen)?;
+
+            if columns
+                .as_ref()
+                .map_or(false, |cols| cols.iter().any(|c| !c.is_column_name()))
+            {
+                return Err(ParserError::new(
+                    column_pos,
+                    "full column definitions are not allowed in CREATE TABLE ... FROM SOURCE",
+                ));
+            }
+
+            Some(TableFromSource {
+                source,
+                external_reference,
+            })
+        } else {
+            // If there is no from_source, we should assert that columns were specified (not None)
+            // and include full definitions
+            match &columns {
+                None => {
+                    return self.expected(
```

Comment:
Checks like this and the diverging parse paths suggest that we may instead need a `CreateTableFromSource` top level struct. Thoughts? Would that make parsing easier to think about and remove the need for this verification?

Reply:
Yes and no. We'd still need many of these checks during parsing since columns are optional in the 'from source' mode, and the column references can just be names rather than both names and types. But it could make downstream logic easier by having distinct `CreateTableStatement` and `CreateTableFromSourceStatement` structs, since I did have to make a couple of fields on `CreateTableStatement` optional to accommodate these semantics.

Reply:
Ok. So that's same parsing but separate structs? That SGTM. Big fan of that if it prevents adding downstream validation that was previously not needed.

Reply:
Yep exactly! Thanks for the feedback, will refactor it when I'm back next week

Reply:
Alright I've replaced this commit with one that introduces a new top-level statement and it does make things dramatically simpler